Disallow interactions by blocked users; some interaction API refactoring (fixes #7971) {Continued in #8755}

CrystallineCat opened the pull request:
Fixes #7971

This PR changes the way objections to user-to-user interactions are handled (e.g. chat revocation or blocks). It includes some refactoring to the API to make it more flexible and allows the UI to query the estimated success of interactions. No UI changes are done at this time.

API refactoring:
- Factors out the checks in `api.sendPrivateMessage` and `api.transferGems` to see if the interaction is allowed.
- The checks are now in `User.getObjectionsToInteractionIfAny`; the method is designed to be easily extensible to new checks or new interactions.
- To allow the UI to check whether an interaction is allowed without actually preforming it, there is a new function `api.getObjectionsToInteractionIfAny`; it only applies interactions initiated by the logged in user and can not be used to query interactions by other users.
- All objections to an interaction now use NotAuthorized instead of NotFound.
- All objection strings are now translatable.

Additional checks:
- Both `api.sendPrivateMessage` and `api.transferGems` now are now blocked by `sender.flags.chatRevoked`.
- `api.transferGems` is now blocked by either user blocking the other.

Tests:
- Tests for the additional checks and the new API have been added.

---

UUID: fea892e5-ac99-4b03-83ac-d01a2154e2af

Commit:
Make flags.chatRevoked prevent sending private messages (issue #7971)

Commit:
Disallow sending gems when messages aren't allowed.

Commit:
Created function to check for objections to an interaction to user model and wired it into the API (issue #7971)

khipkin commented on test/api/v3/integration/members/GET-objections_to_interaction_if_any.test.js:
```diff
@@ -0,0 +1,56 @@
+  translate as t,
+} from '../../../../helpers/api-v3-integration.helper';
+import { v4 as generateUUID } from 'uuid';
+// import common from '../../../../../common';
```
Suggestion: delete this line. :)

khipkin commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+  let knownInteractions = Object.keys(checks).filter((k) => k !== 'always');
+
+  if (!knownInteractions.includes(interaction)) {
+    throw new Error(`Unknown kind of interaction: "${interaction}", expected one of ${knownInteractions.join(', ')}`);
```
Suggestion: add this string to a file in the common/locales/en directory with an "interaction" parameter.  The known interactions can be manually added (probably) because there are a known, finite number of allowed interactions.

(You should verify this with someone on the Habitica staff before implementing though. :smile: )

crookedneighbor replied:
This would be a developer error, not something that should be exposed to the user, so it shouldn't be added to the locale files.

khipkin replied:
In that case, how should it be handled?

crookedneighbor replied:
Just like this. Throwing a normal error.

khipkin commented on test/api/v3/integration/members/POST-transfer_gems.test.js:
```diff
@@ -59,6 +59,48 @@ describe('POST /members/transfer-gems', () => {
   });
 
+  it('returns error when to user has blocked the sender', async () => {
+    let receiverBlocks = await generateUser({'inbox.blocks': [userToSendMessage._id]});
```
Maybe it's just me, but this variable name is confusing me.

khipkin commented on test/api/v3/integration/members/POST-transfer_gems.test.js:
```diff
@@ -59,6 +59,48 @@ describe('POST /members/transfer-gems', () => {
+    });
+  });
+
+  it('returns an error when chat privileges are revoked', async () => {
```
What about if someone sends gems to someone with chat privileges revoked?

Alys replied:
If the recipient has chat permissions revoked, sending gems to them is allowed.

If the sender has site-wide chat privileges revoked, or has been personally blocked by the recipient, then they should not be able to send gems.

Does that cover what you're asking and/or what the code is doing / not doing?  (I haven't checked the code in detail, sorry. :( )

khipkin replied:
I was just suggesting that test coverage for the above situation be added. :smile_cat:

khipkin commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,44 @@ api.getChallengeMemberProgress = {
+    try {
+      response = sender.getObjectionsToInteractionIfAny(req.params.interaction, receiver);
+    } catch (e) {
+      throw new NotFound(e.message);
```
Maybe allow the callee to throw this message ifneedbe? And don't catch it here (in the caller)?

CrystallineCat replied:
The callee is throwing a normal `Error` instead of a HTTP `NotFound` to be independent of HTTP. Unfortunately, an unhandled `Error` will be turned into a generic 500 Internal Server Error and the message lost. This is fine for most API functions, but when the cause of the error can be an API parameter directly, I think the error message needs to be exposed to the user.

khipkin commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
   return userGroups;
 };
 
+// Get an array of error message keys that would be thrown if the given interaction was attempted
```
Technically, if the given interaction was attempted, only one error would be thrown, since it would be pointless to keep trying once a failure condition was found.  Is it really necessary to return ALL issues with an interaction?  Why not just the first one found (if a problem exists) or none (if the interaction is allowed)?

CrystallineCat replied:
I've thought about that, but I'm not sure it's the way to go. I'd rather change the error reporting code to be able to report multiple issues -- the new API function already does that.

Reporting only one when there are multiple seems to me to give a skewed view of the situation, kinda like "fix this one error and it will work", while reporting all gives the full view -- but maybe I'm reading too much into things here.

khipkin commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+      // Private messaging has an opt-out, which does not affect other interactions
+      (sndr, rcvr) => rcvr.inbox.optOut && 'notAuthorizedToSendMessageToThisUser',
+
+      // NB: We allow a player to message themselves so they can test how PMs work or send their own notes to themselves
```
What does NB stand for?

Alys replied:
Nota Bene ("note well"). Let's remove "NB" since it's not necessary and few of us are ancient Romans. :) Sorry, CrystallineCat - entirely my fault. I hadn't thought that it wasn't widely known.

CrystallineCat replied:
Actually, it's not your fault, the NB wasn't in your original line. I added it because the comment looked a bit out of place to me without any code to follow it. I didn't consider it might confuse people, so I agree it's better to remove it. :)

Alys replied:
Oh! Huh. :) "NB" is something I use myself so often I just assumed I'd specified that it should be used. :)

crookedneighbor commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+schema.methods.getObjectionsToInteractionIfAny = function getObjectionsToInteractionIfAny (interaction, receiver) {
+  let sender = this;
+
+  /* eslint-disable no-unused-vars */
```
Why disable eslint here?

CrystallineCat replied:
The checks are all of the form `(sndr, rcvr) => ...`, due to being iterated over and called with the same signature, but not all of them use both parameters. This leads to spurious `no-unused-vars` issues.

I think following the linter here would be wrong and make the code worse, not better. I've added a comment stating the reason for the disable.

crookedneighbor commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
 };
 
+// Get an array of error message keys that would be thrown if the given interaction was attempted
+schema.methods.getObjectionsToInteractionIfAny = function getObjectionsToInteractionIfAny (interaction, receiver) {
```
Can we remove the `IfAny` from the name? I don't think that's necessary.

crookedneighbor commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+  };
+  /* eslint-enable no-unused-vars */
+
+  let knownInteractions = Object.keys(checks).filter((k) => k !== 'always');
```
Lets avoid one character variables. `filter(check => check !== 'aways');`  should do it.

Commit:
Fixes for issues raised by reviewers.

crookedneighbor commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,45 @@ api.getChallengeMemberProgress = {
+ * @apiGroup Member
+ *
+ * @apiParam {UUID} toUserId The user to interact with
+ * @apiParam {String} interaction Name of the interaction to query, e.g. "send-private-message"
```
We should enumerate in the API docs all the interactions that can be queried.

Commit:
Added allowed values to apidoc for api.getObjectionsToInteraction.

CrystallineCat commented:
@Alys @crookedneighbor @khipkin: Is there anything else I need to do to get this merged? I think I addressed all the comments. Apologies if I missed anything.

crookedneighbor commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,45 @@ api.getChallengeMemberProgress = {
+ */
+api.getObjectionsToInteraction = {
+  method: 'GET',
+  url: '/members/:toUserId/objections-to/:interaction',
```
I'm not thrilled about the name of this url (the `objections-to` part), but I can't think of a better name :(

crookedneighbor replied:
Can we mark this route as beta so we can change it without worrying about breaking 3rd party integrations?

CrystallineCat replied:
I added "- BETA" to the comment; is anything else required to mark it as beta?

TheHollidayInn replied:
Why don't we just called the route `objections`? You are getting objections right?

crookedneighbor commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,45 @@ api.getChallengeMemberProgress = {
 };
 
 /**
+ * @api {get} /api/v3/members/:toUserId/objections-to/:interaction Get the message of any errors that would occur if the given interaction was attempted
```
Can we fill out this documentation a bit? See https://github.com/HabitRPG/habitrpg/pull/7879/files#diff-cf049c3d45b117fee868e17845f1f130R8 for an example.

CrystallineCat replied:
I'm not sure there is anything to add. The example is of a method with a much more complicated parameter structure; the current documentation for the new method already includes everything about its parameters. The only missing information is the exception, and I don't think committing to a specific exception here is a good idea.

crookedneighbor commented on test/api/v3/integration/members/GET-objections_to_interaction.test.js:
```diff
@@ -0,0 +1,55 @@
+    ).to.eventually.be.rejected.and.eql({
+      code: 404,
+      error: 'NotFound',
+      message: 'Unknown kind of interaction: "hug-a-whole-forest-of-trees", expected one of send-private-message, transfer-gems',
```
This should be a translation string with interpolations:

``` json
"uknownInteraction": "Uknown kind of interaction: \"<%= unknownInteraction %>\". Expected one of <%= interactions %>"
```

``` js
let unknownInteraction = req.params.interaction;
let interactions = 'send-private-message, transfer-gems';
t('uknownInteraction', { unknownInteraction, interactions });
```

crookedneighbor replied:
If using my thought to export the known interactions:

``` js
t('uknownInteraction', { unknownInteraction, interactions: KNOWN_INTERACTIONS.join(',') });
```

CrystallineCat replied:
I decided to move this to the parameter checking with req.checkParams. This provides a less useful error message, but is consistent with the other methods and I'm feeling uneasy about handling this check separately. Ideally, all the parameter checking would be done data-driven, but that's out of scope for this, I think.

crookedneighbor commented on test/api/v3/integration/members/GET-objections_to_interaction.test.js:
```diff
@@ -0,0 +1,55 @@
+    });
+  });
+
+  it('returns nothing if there are no objections', async () => {
```
Technically it returns an empty array, not nothing.

crookedneighbor commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,45 @@ api.getChallengeMemberProgress = {
+      response = sender.getObjectionsToInteraction(req.params.interaction, receiver);
+    } catch (e) {
+      // Rethrow, so that the message gets passed to the client
+      throw new NotFound(e.message);
```
I feel weird about this. Feels like the check for valid params should happen much earlier, before we look up the receiver.

crookedneighbor commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+  };
+  /* eslint-enable no-unused-vars */
+
+  let knownInteractions = Object.keys(checks).filter((key) => key !== 'always');
```
The `checks` and `knownInteractions` variables can be pull out of this function since they are constant and do not need to be recreated for every time this method is called. 

Maybe:

``` js
const INTERACTION_CHECKS = Object.freeze({ /* functions here */ });
const KNOWN_INTERACTIONS = Object.freeze(Object.keys(INTERACTION_CHECKS).filter(key => key !== 'always'));

This means you can also export `KNOWN_INTERACTIONS` for a params check in the route instead of throwing an error here.
```

crookedneighbor commented on website/server/models/user/methods.js:
```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
+    throw new Error(`Unknown kind of interaction: "${interaction}", expected one of ${knownInteractions.join(', ')}`);
+  }
+
+  let checksToRun = checks.always.concat(checks[interaction]);
```
Might be better to use lodash chaining for the following lines so we're not creating a ton of extra arrays

crookedneighbor replied:
^ How is this going?

CrystallineCat replied:
I've converted it to lodash. I'm fairly new to JS and haven't used that library before, so I hope I did it the right way. :)

TBH, I'm not sure I'm that concerned about the arrays; the number is constant per call and I don't think the method is going to be a bottleneck, as it is likely only to be called in response to relatively infrequent UI events.

I realize I've gone against some of your requests, I've considered them over and over again and couldn't get myself to agree with them. Apologies if this causes any issues and thank you for your reviews. Also apologies for the huge delay... it took me being shocked from recent events to numb my fear of this conversation enough... :/ I'll gladly step back if you'd prefer someone else to work on this.

crookedneighbor commented on website/server/controllers/api-v3/members.js:
```diff
@@ -270,6 +270,45 @@ api.getChallengeMemberProgress = {
+ * @apiParam {UUID} toUserId The user to interact with
+ * @apiParam {String="send-private-message","transfer-gems"} interaction Name of the interaction to query
+ *
+ * @apiSuccess {Array} data Return an array of error messages, if the interaction would be blocked; otherwise an empty array
```
I think it should return an array of objection messages, rather than error messages.

CrystallineCat replied:
I'm not sure what you mean by "objection messages".

TheHollidayInn replied:
I think his is saying that since this route returns a list of objects, you should change your wroding to say "Return an array of objections"

Commit:
Refactoring of getObjectionsToInteraction and minor API changes.

TheHollidayInn commented on test/api/v3/integration/members/GET-objections_to_interaction.test.js:
```diff
@@ -0,0 +1,54 @@
+      user.get(`/members/${receiver._id}/objections-to/send-private-message`)
+    ).to.eventually.be.fulfilled.and.eql([]);
+  });
+});
```
You should added a test here to show that the route returns objections

paglias commented on website/server/models/user/methods.js:
```diff
@@ -4,8 +4,7 @@ import {
 } from '../group';
-import { defaults } from 'lodash';
-
+import { defaults, map, flatten, flow, compact, unique, partialRight } from 'lodash';
```
Heads up, we're migrating to lodash v4 (https://github.com/HabitRPG/habitica/pull/8495) which includes many breaking changes (https://github.com/lodash/lodash/wiki/Changelog#v400). Before merging this PR should be checked and updated to make sure all lodash methods are compatible with v4.

SabreCat commented:
Continued in #8755!
